File: tests/test_init.py
"""Tests for 'corrkit init' command."""

import tomllib

import pytest

import app_config
from init import main


@pytest.fixture(autouse=True)
def _isolate_app_config(tmp_path, monkeypatch):
    """Isolate app_config to a temp directory so tests don't write to real config."""
    config_dir = tmp_path / "app-config"
    config_dir.mkdir()
    monkeypatch.setattr(app_config, "app_config_dir", lambda: config_dir)


def test_init_creates_directory_structure(tmp_path, monkeypatch):
    """init creates conversations/, drafts/, contacts/ with .gitkeep."""
    data_dir = tmp_path / "correspondence"
    monkeypatch.setattr(
        "sys.argv",
        ["init", "--user", "test@example.com", "--data-dir", str(data_dir)],
    )

    main()

    for sub in ("conversations", "drafts", "contacts"):
        assert (data_dir / sub).is_dir()
        assert (data_dir / sub / ".gitkeep").exists()


def test_init_creates_accounts_toml(tmp_path, monkeypatch):
    """init generates valid accounts.toml with correct provider/user/labels."""
    data_dir = tmp_path / "correspondence"
    monkeypatch.setattr(
        "sys.argv",
        [
            "init",
            "--user",
            "test@gmail.com",
            "--data-dir",
            str(data_dir),
            "--provider",
            "gmail",
            "--labels",
            "inbox,sent",
            "--github-user",
            "testuser",
            "--name",
            "Test User",
        ],
    )

    main()

    accounts_path = data_dir / "accounts.toml"
    assert accounts_path.exists()

    with open(accounts_path, "rb") as f:
        raw = tomllib.load(f)

    assert raw["owner"]["github_user"] == "testuser"
    assert raw["owner"]["name"] == "Test User"
    assert raw["accounts"]["default"]["provider"] == "gmail"
    assert raw["accounts"]["default"]["user"] == "test@gmail.com"
    assert raw["accounts"]["default"]["labels"] == ["inbox", "sent"]
    assert raw["accounts"]["default"]["default"] is True


def test_init_creates_empty_config_files(tmp_path, monkeypatch):
    """init creates empty collaborators.toml and contacts.toml."""
    data_dir = tmp_path / "correspondence"
    monkeypatch.setattr(
        "sys.argv",
        ["init", "--user", "test@example.com", "--data-dir", str(data_dir)],
    )

    main()

    assert (data_dir / "collaborators.toml").exists()
    assert (data_dir / "contacts.toml").exists()


def test_init_force_overwrites(tmp_path, monkeypatch):
    """--force overwrites existing accounts.toml."""
    data_dir = tmp_path / "correspondence"
    data_dir.mkdir(parents=True)
    accounts = data_dir / "accounts.toml"
    accounts.write_text("old content", encoding="utf-8")

    monkeypatch.setattr(
        "sys.argv",
        [
            "init",
            "--user",
            "new@example.com",
            "--data-dir",
            str(data_dir),
            "--force",
        ],
    )

    main()

    content = accounts.read_text(encoding="utf-8")
    assert "new@example.com" in content
    assert "old content" not in content


def test_init_without_force_exits_if_exists(tmp_path, monkeypatch):
    """Without --force, init exits 1 if accounts.toml already exists."""
    data_dir = tmp_path / "correspondence"
    data_dir.mkdir(parents=True)
    (data_dir / "accounts.toml").write_text("existing", encoding="utf-8")

    monkeypatch.setattr(
        "sys.argv",
        ["init", "--user", "test@example.com", "--data-dir", str(data_dir)],
    )

    with pytest.raises(SystemExit) as exc_info:
        main()
    assert exc_info.value.code == 1


def test_init_password_cmd_in_toml(tmp_path, monkeypatch):
    """--password-cmd is written to accounts.toml."""
    data_dir = tmp_path / "correspondence"
    monkeypatch.setattr(
        "sys.argv",
        [
            "init",
            "--user",
            "test@example.com",
            "--data-dir",
            str(data_dir),
            "--password-cmd",
            "pass email/test",
        ],
    )

    main()

    with open(data_dir / "accounts.toml", "rb") as f:
        raw = tomllib.load(f)
    assert raw["accounts"]["default"]["password_cmd"] == "pass email/test"


def test_init_protonmail_provider(tmp_path, monkeypatch):
    """init with protonmail-bridge provider works."""
    data_dir = tmp_path / "correspondence"
    monkeypatch.setattr(
        "sys.argv",
        [
            "init",
            "--user",
            "test@pm.me",
            "--data-dir",
            str(data_dir),
            "--provider",
            "protonmail-bridge",
            "--password-cmd",
            "pass email/proton",
        ],
    )

    main()

    with open(data_dir / "accounts.toml", "rb") as f:
        raw = tomllib.load(f)
    assert raw["accounts"]["default"]["provider"] == "protonmail-bridge"


def test_init_registers_space(tmp_path, monkeypatch):
    """init registers the data dir as a space in app config."""
    data_dir = tmp_path / "correspondence"
    monkeypatch.setattr(
        "sys.argv",
        ["init", "--user", "test@example.com", "--data-dir", str(data_dir)],
    )

    main()

    config = app_config.load()
    assert "default" in config["spaces"]
    assert config["spaces"]["default"]["path"] == str(data_dir)
    assert config["default_space"] == "default"


def test_init_custom_space_name(tmp_path, monkeypatch):
    """--space sets a custom space name."""
    data_dir = tmp_path / "correspondence"
    monkeypatch.setattr(
        "sys.argv",
        [
            "init",
            "--user",
            "test@example.com",
            "--data-dir",
            str(data_dir),
            "--space",
            "work",
        ],
    )

    main()

    config = app_config.load()
    assert "work" in config["spaces"]
    assert config["default_space"] == "work"

File: tests/test_init.rs
//! Integration tests for corky init (src/init.rs).
//!
//! Each test sets HOME to a temp dir to isolate from the real
//! ~/.config/corky/config.toml. Tests run serially via
//! a shared mutex to avoid env var races.

mod common;

use std::sync::Mutex;
use tempfile::TempDir;

use corky::accounts::{load_accounts, load_owner};

static ENV_MUTEX: Mutex<()> = Mutex::new(());

/// Run init::run with HOME set to the temp dir parent, so
/// app_config::add_mailbox writes to an isolated config.
#[allow(clippy::too_many_arguments)]
fn run_init_isolated(
    tmp: &TempDir,
    path: &std::path::Path,
    user: &str,
    provider: &str,
    password_cmd: &str,
    labels: &str,
    github_user: &str,
    name: &str,
    mailbox: &str,
    force: bool,
) -> anyhow::Result<()> {
    let _lock = ENV_MUTEX.lock().unwrap();
    let old_home = std::env::var("HOME").ok();
    std::env::set_var("HOME", tmp.path().to_string_lossy().as_ref());
    let result = corky::init::run(
        user, path, provider, password_cmd, labels, github_user, name,
        false, // sync
        mailbox, force,
        false, // with_skill
    );
    // Restore HOME
    if let Some(h) = old_home {
        std::env::set_var("HOME", h);
    }
    result
}

#[test]
fn test_init_creates_directory_structure() {
    let tmp = TempDir::new().unwrap();
    let path = tmp.path().join("myproject");

    run_init_isolated(
        &tmp, &path, "test@example.com", "gmail", "",
        "correspondence", "testgh", "Test User",
        "test-init-mb", true,
    )
    .unwrap();

    let data_dir = path.join("mail");
    assert!(data_dir.join("conversations").is_dir());
    assert!(data_dir.join("drafts").is_dir());
    assert!(data_dir.join("contacts").is_dir());
    assert!(data_dir.join("conversations").join(".gitkeep").exists());
    assert!(data_dir.join("drafts").join(".gitkeep").exists());
    assert!(data_dir.join("contacts").join(".gitkeep").exists());
    // Config inside mail/
    assert!(data_dir.join(".corky.toml").exists());
    assert!(data_dir.join("contacts.toml").exists());
    assert!(data_dir.join("voice.md").exists());
    // No config at project root
    assert!(!path.join(".corky.toml").exists());
    assert!(!path.join("contacts.toml").exists());
    assert!(!path.join("voice.md").exists());
}

#[test]
fn test_init_corky_toml_content() {
    let tmp = TempDir::new().unwrap();
    let path = tmp.path().join("initdata");

    run_init_isolated(
        &tmp, &path, "alice@gmail.com", "gmail",
        "pass show email/personal", "inbox, sent",
        "alicegh", "Alice", "test-init-mb-acct", true,
    )
    .unwrap();

    let config_path = path.join("mail").join(".corky.toml");
    let accounts = load_accounts(Some(&config_path)).unwrap();
    assert!(accounts.contains_key("default"));
    let acct = accounts.get("default").unwrap();
    assert_eq!(acct.provider, "gmail");
    assert_eq!(acct.user, "alice@gmail.com");
    assert!(acct.default);

    let owner = load_owner(Some(&config_path)).unwrap();
    assert_eq!(owner.github_user, "alicegh");
    assert_eq!(owner.name, "Alice");
}

#[test]
fn test_init_with_custom_provider() {
    let tmp = TempDir::new().unwrap();
    let path = tmp.path().join("pmdata");

    run_init_isolated(
        &tmp, &path, "user@proton.me", "protonmail-bridge",
        "", "correspondence", "", "", "test-init-mb-pm", true,
    )
    .unwrap();

    let config_path = path.join("mail").join(".corky.toml");
    let accounts = load_accounts(Some(&config_path)).unwrap();
    let acct = accounts.get("default").unwrap();
    assert_eq!(acct.provider, "protonmail-bridge");
    assert_eq!(acct.user, "user@proton.me");
}

#[test]
fn test_init_labels_parsing() {
    let tmp = TempDir::new().unwrap();
    let path = tmp.path().join("lbldata");

    run_init_isolated(
        &tmp, &path, "user@example.com", "imap",
        "", "inbox, sent, important", "", "",
        "test-init-mb-labels", true,
    )
    .unwrap();

    let config_path = path.join("mail").join(".corky.toml");
    let accounts = load_accounts(Some(&config_path)).unwrap();
    let acct = accounts.get("default").unwrap();
    assert_eq!(acct.labels.len(), 3);
    assert!(acct.labels.contains(&"inbox".to_string()));
    assert!(acct.labels.contains(&"sent".to_string()));
    assert!(acct.labels.contains(&"important".to_string()));
}

#[test]
fn test_init_force_overwrites() {
    let tmp = TempDir::new().unwrap();
    let path = tmp.path().join("forcedata");
    let data_dir = path.join("mail");
    std::fs::create_dir_all(&data_dir).unwrap();
    std::fs::write(data_dir.join(".corky.toml"), "# old config").unwrap();

    run_init_isolated(
        &tmp, &path, "new@example.com", "gmail",
        "", "correspondence", "", "",
        "test-init-mb-force", true,
    )
    .unwrap();

    let content = std::fs::read_to_string(data_dir.join(".corky.toml")).unwrap();
    assert!(content.contains("new@example.com"));
    assert!(!content.contains("# old config"));
}

#[test]
fn test_init_tilde_expansion() {
    let tmp = TempDir::new().unwrap();
    let path = tmp.path().join("tildetest");

    run_init_isolated(
        &tmp, &path, "user@example.com", "gmail",
        "", "correspondence", "", "",
        "test-init-mb-tilde", true,
    )
    .unwrap();

    assert!(path.join("mail").join(".corky.toml").exists());
}

#[test]
fn test_init_empty_labels() {
    let tmp = TempDir::new().unwrap();
    let path = tmp.path().join("emptylbl");

    run_init_isolated(
        &tmp, &path, "user@example.com", "gmail",
        "", "", "", "",
        "test-init-mb-emptylbl", true,
    )
    .unwrap();

    let config_path = path.join("mail").join(".corky.toml");
    let accounts = load_accounts(Some(&config_path)).unwrap();
    let acct = accounts.get("default").unwrap();
    assert!(acct.labels.is_empty());
}

#[test]
fn test_init_gitignore_in_git_repo() {
    let tmp = TempDir::new().unwrap();
    let path = tmp.path().join("gitproject");
    std::fs::create_dir_all(&path).unwrap();
    // Simulate a git repo
    std::fs::create_dir_all(path.join(".git")).unwrap();

    run_init_isolated(
        &tmp, &path, "user@example.com", "gmail",
        "", "correspondence", "", "",
        "test-init-mb-gitignore", true,
    )
    .unwrap();

    let gitignore = std::fs::read_to_string(path.join(".gitignore")).unwrap();
    assert!(gitignore.contains("mail"));
}

#[test]
fn test_init_no_gitignore_without_git() {
    let tmp = TempDir::new().unwrap();
    let path = tmp.path().join("nogit");

    run_init_isolated(
        &tmp, &path, "user@example.com", "gmail",
        "", "correspondence", "", "",
        "test-init-mb-nogit", true,
    )
    .unwrap();

    assert!(!path.join(".gitignore").exists());
}

#[test]
fn test_init_with_skill() {
    let tmp = TempDir::new().unwrap();
    let path = tmp.path().join("skillproject");

    let _lock = ENV_MUTEX.lock().unwrap();
    let old_home = std::env::var("HOME").ok();
    std::env::set_var("HOME", tmp.path().to_string_lossy().as_ref());
    let result = corky::init::run(
        "user@example.com", &path, "gmail", "", "correspondence", "", "",
        false, "test-init-mb-skill", true,
        true, // with_skill
    );
    if let Some(h) = old_home {
        std::env::set_var("HOME", h);
    }
    result.unwrap();

    assert!(path.join(".claude/skills/email/SKILL.md").exists());
    assert!(path.join(".claude/skills/email/README.md").exists());
}
